.circleci/config.yml

```diff
@@ -584,6 +589,49 @@ commands:
             cd << parameters.repo_path >>/study-builder
             mvn test-compile --batch-mode -DskipTests -Dmaven.repo.local=<< parameters.m2_repo >>
 
+  restore-ddp-cache:
+    description: Restore ddp cache
+    steps:
+      - run:
+          name: Download ddp-cache-md5 artifact
+          command: |
+            ddp_cache_md5='/home/circleci/.ddp-cache-md5.txt'
+            latest_artifact_url="https://circleci.com/api/v1.1/project/github/broadinstitute/ddp-study-server/latest/artifacts?circle-token=$CIRCLE_TOKEN"
+            md5_artifact_url=$(curl -sL "$latest_artifact_url" | jq -r '.[] | select(.node_index == 0) | select(.path == "ddp-cache-md5.txt") | .url')
+            if [[ -n "$md5_artifact_url" ]]; then
+              echo "Downloading artifact file [$md5_artifact_url] to file [$ddp_cache_md5]"
+              curl -sL "$md5_artifact_url?circle-token=$CIRCLE_TOKEN" -o "$ddp_cache_md5"
+            else
+              echo "No artifact file found, creating empty file [$ddp_cache_md5]"
+              touch "$ddp_cache_md5"
+            fi
+      - restore_cache:
+          keys:
+            - ddp-cache-{{ checksum "/home/circleci/.ddp-cache-md5.txt" }}
+            - ddp-cache-
+
+  save-ddp-cache:
+    description: Save ddp cache
+    steps:
+      - run:
+          name: Compute ddp-cache-md5
+          command: |
+            ddp_cache_dir='/home/circleci/.ddp-cache'
+            ddp_cache_md5='/home/circleci/.ddp-cache-md5.txt'
+            echo -n "" > "$ddp_cache_md5"
+            for cached_file in $(ls "$ddp_cache_dir" | sort); do
+              md5sum "$ddp_cache_dir/$cached_file" >> "$ddp_cache_md5"
+            done
+            echo "Contents of [$ddp_cache_md5]:"
+            cat "$ddp_cache_md5"
+      - save_cache:
+          key: ddp-cache-{{ checksum "/home/circleci/.ddp-cache-md5.txt" }}
+          paths:
+            - /home/circleci/.ddp-cache
+      - store_artifacts:
+          path: /home/circleci/.ddp-cache-md5.txt
+          destination: ddp-cache-md5.txt
+
```

Comment:
Not getting how the md5s are helping us here. Wonder if just simpler to save_cache just like you have it here and using the TestingUserUtil.java to generate the key suffix checksum. You already have the retry when we rate-limited.

Reply:
I guess my concern here is that it is not clear to me that the keys you are generating would be able to be recreated in follow up executions of the script, in which case they can't be retrieved. Looks like you have resolved this problem by storing as artifacts. Why not just use the artifacts?
Regardless, from your code looks to me that the Auth0 retries with waits will work no matter what. Did you. consider just ditching the caching/storing altogether? Shorter scripts/easier to grokk? (Question, not request).

Reply:
The trouble I have with artifacts is that they seem to get overwritten / cleared out if the latest build fails. This means the next build that comes along will not find any artifacts. This is also the reason why my solution here is not optimal since it relies on artifact for the cache key. But at the very least it's good enough for now. Also, the benefit of keeping with cache is, unlike artifacts, there is _something_ cached, so you can retrieve it, albeit it being a bit older.

The trouble with the Auth0 retries is that I have only put it in one place. I was not able to track down the other places where we're hitting the rate limit yet. Regardless, to answer your question, I think it's better to cache things if we can instead of needlessly always calling Auth0 for a token.